src/prefect/backend/task_run.py

```diff
@@ -0,0 +1,233 @@
+                    order_by={"map_index": EnumValue("asc")},
+                    many=True,
+                )
+            ]
+
+            self.state.map_states = [task_run.state for task_run in child_task_runs]
+
+        # Fire the state result hydration
+        self.state.load_result()
+        return self.state.result
+
+    @classmethod
+    def from_task_run_data(cls, task_run: dict) -> "TaskRun":
+        task_run_id = task_run.pop("id")
+        task_data = task_run.pop("task")
+        serialized_state = task_run.pop("serialized_state")
+
+        return cls(
+            task_run_id=task_run_id,
+            state=State.deserialize(serialized_state),
+            task_id=task_data["id"],
+            task_slug=task_data["slug"],
+            **task_run,
+        )
+
+    @classmethod
+    def from_task_run_id(cls, task_run_id: str = None) -> "TaskRun":
+        if not isinstance(task_run_id, str):
+            raise TypeError(
+                f"Unexpected type {type(task_run_id)!r} for `task_run_id`, "
+                f"expected 'str'."
+            )
+
+        return cls.from_task_run_data(
+            cls.query_for_task_runs(where={"id": {"_eq": task_run_id}})
+        )
+
+    @classmethod
+    def from_task_slug(cls, task_slug: str, flow_run_id: str) -> "TaskRun":
+        return cls.from_task_run_data(
+            cls.query_for_task_runs(
+                where={
+                    "task": {"slug": {"_eq": task_slug}},
+                    "flow_run_id": {"_eq": flow_run_id},
+                    # Since task slugs can be duplicated for mapped tasks, only allow
+                    # the root task to be pulled by this
+                    "map_index": {"_eq": -1},
+                }
+            )
+        )
+
+    @staticmethod
+    def query_for_task_runs(
+        where: dict,
+        many: bool = False,
+        order_by: dict = None,
+        error_on_empty: bool = True,
+    ) -> Union[dict, List[dict]]:
+        """
+        Query for task run data necessary to initialize `TaskRun` instances
+        with `TaskRun.from_task_run_data`.
+
+        Args:
+            where (required): The Hasura `where` clause to filter by
+            many (optional): Are many results expected? If `False`, a single record will
+                be returned and if many are found by the `where` clause an exception
+                will be thrown. If `True` a list of records will be returned.
+            order_by (optional): An optional Hasura `order_by` clause to order results
+                by. Only applicable when `many` is `True`
+            error_on_empty (optional): If `True` and no tasks are found, a `ValueError`
+                will be raised. If `False`, an empty list or dict will be returned
+                based on the value of `many`.
+
+        Returns:
+            A dict of task run information (or a list of dicts if `many` is `True`)
+        """
```

Comment:
```python
Raises
  - ...
  ```

Reply:
I'm not going to add these until things are more finalized, but that's a good note